src/System.Net.Http/src/System/Net/Http/Managed/AuthenticateAndRedirectHandler.cs

```diff
@@ -2,73 +2,59 @@
+        public AuthenticateAndRedirectHandler(bool preAuthenticate, ICredentials credentials, bool allowRedirect, int maxAutomaticRedirections, HttpMessageHandler innerHandler)
         {
-            _innerHandler = innerHandler ?? throw new ArgumentNullException(nameof(innerHandler));
+            Debug.Assert(innerHandler != null);
 
-            if (maxAutomaticRedirections < 0)
-            {
-                throw new ArgumentOutOfRangeException(nameof(maxAutomaticRedirections));
-            }
-            _maxAutomaticRedirections = maxAutomaticRedirections;
-        }
-
-        internal static bool RequestNeedsRedirect(HttpResponseMessage response)
-        {
-            bool needRedirect = false;
-            switch (response.StatusCode)
-            {
-                case HttpStatusCode.Moved:
-                case HttpStatusCode.Found:
-                case HttpStatusCode.SeeOther:
-                case HttpStatusCode.TemporaryRedirect:
-                    needRedirect = true;
-                    break;
-
-                case HttpStatusCode.MultipleChoices:
-                    needRedirect = response.Headers.Location != null; // Don't redirect if no Location specified
-                    break;
-            }
+            _preAuthenticate = preAuthenticate;
+            _credentials = credentials;
+            _allowRedirect = allowRedirect;
 
-            return needRedirect;
-        }
-
-        private static bool RequestRequiresForceGet(HttpStatusCode statusCode, HttpMethod requestMethod)
-        {
-            if (statusCode == HttpStatusCode.Moved ||
-                statusCode == HttpStatusCode.Found ||
-                statusCode == HttpStatusCode.SeeOther ||
-                statusCode == HttpStatusCode.MultipleChoices)
+            if (allowRedirect && maxAutomaticRedirections < 0)
             {
-                return requestMethod == HttpMethod.Post;
+                throw new ArgumentOutOfRangeException(nameof(maxAutomaticRedirections));
             }
 
-            return false;
+            _maxAutomaticRedirections = maxAutomaticRedirections;
+            _innerHandler = innerHandler;
         }
 
         protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         {
             HttpResponseMessage response;
             uint redirectCount = 0;
+            bool useCredentialCache = false;
             while (true)
             {
-                response = await _innerHandler.SendAsync(request, cancellationToken).ConfigureAwait(false);
+                response = await SendRequestAsync(request, useCredentialCache, cancellationToken).ConfigureAwait(false);
 
                 if (!RequestNeedsRedirect(response))
                 {
                     break;
                 }
 
+                // Clear the authorization header, if the request requires redirect.
+                request.Headers.Authorization = null;
```

Comment:
Is this actually mutating the original HttpRequestMessage that a dev would have passed into SendAsync()?  If so, we shouldn't be changing that.

Reply:
Why not? We modify the request object in all of the managed handlers, including adding authorization headers, changing request method/location on redirection, adding cookie headers, etc.

Reply:
>We modify the request object in all of the managed handlers, including adding authorization headers, changing request method/location on redirection, adding cookie headers, etc.

How many managed handlers are there?  I thought there was only one?

Other than modifying HttpRequestMessage.RequestUri (to update it to a redirected URI), the original request message isn't normally changed.

The fact that the managed handler has been doing that doesn't mean that it is the right thing to do.  The .NET Framework, .NET Core (WinHttpHandler), .NET Core (UWP) doesn't change the original request message.

I don't understand the design philosophy behind the "managed handler" (i.e. why is it designed as a chain of 'handler's).

@stephentoub can you shed some light on all of this?  This design of the managed handler doesn't seem to align with the original design principles of HttpClient et. al.

Reply:
> How many managed handlers are there? I thought there was only one?

Currently without this change, we have, AuthenticationHandler, AutoRedirectHandler, CookieHandler, DecompressionHandler, HttpConnectionHandler, HttpProxyConnectionHandler.

Reply:
The original design of handlers and having a chain wasn't intended for the protocol stack itself to be implemented as a chain of handlers.  So, this really goes back to the original design concepts of how this managed handler came to be.  Why isn't it just a single HANDLER? Is there some design document that explains the motivations for this design?

Reply:
I am not aware of any documentation for the design - I believe the motivation is componentization. Note that @Tratcher pointed out it will be likely broken when we implement authentication.
While handlers may not have been designed originally for this purpose, is there any specific concern for using them this way @davidsh?

Reply:
I think might have been better ways to achieve componentization in implementing this protocol stack besides using the pipeline model that was intended only for upstack components.

And in terms of concerns, I've raised them already including the fact that the "handler" shouldn't be modifying all those field/properties of the original HttpRequestMessage.

Reply:
Folks, we're bikeshedding here.  This layering inside the ManagedHandler is purely implementation detail; if there are better ways for this or that to be implemented, it can be restructured.  We've already talked about multiple reasons why we might want to collapse some of this layering, e.g. minimizing extra allocations, making it easier to flow certain kinds of auth-related data, etc., and I suggested to Priya offline that collapsing the auth and redirect pieces would make this easier and more efficient to implement as well (simply passing this auth state around as a "local" in the state machine for the collapsed layers).  Let's stop hashing on that layering here.

As for not modifying the HttpRequestMessage, @davidsh you wrote "This design of the managed handler doesn't seem to align with the original design principles of HttpClient et. al.".  Can you please point to the documentation for those design principles?  WinHttpHandler even messes with the HttpRequestMessage in multiple places, e.g.
https://github.com/dotnet/corefx/blob/master/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs#L587
https://github.com/dotnet/corefx/blob/master/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs#L600
https://github.com/dotnet/corefx/blob/master/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs#L197
and of course HttpClient itself messes with it, e.g.
https://github.com/dotnet/corefx/blob/6bf1e7c3032d3c24d5a41bbd880549ef4a8de2da/src/System.Net.Http/src/System/Net/Http/HttpClient.cs#L650
so obviously changes are made beyond those the dev makes.